### ISSUE TITLE
Windows debug test failure introduced around Sept 16 2017

### ISSUE THREAD

**Issue by dhood** [ros2/build_farmer#59]
http://ci.ros2.org/view/nightly/job/nightly_win_deb/609/ was the first job that had the following failure:

```
test_publisher_subscriber__DynamicArrayPrimitives__rclcpp__rclpy__rmw_connext_cpp__rmw_fastrtps_cpp_Debug.test_publisher_subscriber
```

and it has failed in most of the windows debug jobs since (but not all). Note that that job is not looking for flaky tests, it uses `--retest-until-pass 10`.

The nature of the failure is that the receiver doesn't receive all of its messages but it usually receives 4 of 5.

PRs from around that time: https://github.com/search?utf8=%E2%9C%93&q=user%3Aament+user%3Aros2+merged%3A%222017-09-14T22%3A00%3A00-08%3A00+..+2017-09-16T23%3A30%3A00-08%3A00%22&type=Issues

haven't narrowed in on what caused it yet.

**Comment by mikaelarguedas** [comment]
hmmm I would expect the cause to be https://github.com/ros2/rcl_interfaces/commit/558cbce984405999f25568b1efbcd6803574583b given that this makes the 5th message MUUUCH longer (but the commit is much older than the date you mentioned).

Is it always the same test ? (c++ connext publisher and python fastrtps subscriber) or does it vary bu always on the same message type ?

**Comment by mikaelarguedas** [comment]
Oh no my bad, thats the 5th message that is not received not the 4th. Sorry for the noise
Reason for this specific test to [be added](https://github.com/ros2/system_tests/pull/118) in the first place: https://github.com/ros2/rclcpp/issues/219.

**Comment by dhood** [comment]
it's always that specific test for some reason. Thanks for the extra context. I'm suspicious of connext5.3 at the moment since that change happened around the same time (didn't see any commits to fastrtps around then)

**Comment by dhood** [comment]
The test had to be re-tested to pass even with 5.2.3: http://ci.ros2.org/job/ci_windows/3340/consoleFull

Historically it has always sometimes taken a while for the subscriber to connect, so I don't think that that's a problem in itself (e.g. [this job in June](http://ci.ros2.org/view/nightly/job/nightly_win_deb/517/testReport/(root)/test_publisher_subscriber__DynamicArrayPrimitives__rclcpp__rclpy__rmw_connext_cpp__rmw_fastrtps_cpp_Debug/test_publisher_subscriber/), or [this job in August](http://ci.ros2.org/view/nightly/job/nightly_win_deb/575/testReport/(root)/test_publisher_subscriber__DynamicArrayPrimitives__rclcpp__rclpy__rmw_connext_cpp__rmw_fastrtps_cpp_Debug/test_publisher_subscriber/))

http://ci.ros2.org/view/nightly/job/nightly_win_deb/603/consoleFull was the first job that had to repeat the test to get it to pass. That ran early morning 11 Sept, so I'm changing the merge window to the day of the 10th: https://github.com/search?utf8=%E2%9C%93&q=user%3Aament+user%3Aros2+merged%3A%222017-09-09T22%3A00%3A00-08%3A00+..+2017-09-10T23%3A30%3A00-08%3A00%22&type=Issues

Not seeing any fastrtps commits for that day, it looks like https://github.com/ros2/rclpy/pull/104 was the only core PR merged that day, so that's what I'm determining as the cause ( @sloretz FYI).
I suppose the question is why the subscriber always starts late now for windows debug, and why can't it ever receive the 5th message. It might be that the way we test that use case has to be updated, or that this is exposing a regression in the code.

**Comment by sloretz** [comment]
I've looked into this a little bit. Sorry for the wall of text. Options are to optimize code to run faster on windows debug, or change the publisher rate slightly on the c++ publisher.

Hypotheses:
* The long time (3 seconds-ish) between the python subscriber message "beginning loop" and the python subscriber printing that it RX'd a message is not in rclpy. It maybe an issue with connext/rclcpp.
* The actual test failure may be slow performance of rclpy executors in debug mode on windows combined with interactions between the QoS profile and the slowness of the subscriber being very close to half the rate the publisher is publishing.

Reasoning:
* Delay between publisher publishing and subscriber receiving the first message
  * When using only rclpy as the publisher and subscriber I never saw a delay more than 1 second, and it was almost always near zero delay. It is likely that the python subscriber is blocked by rcl_wait during this time. When the publisher is rclcpp/connext there is almost always a consistent 3 second delay. When the subscriber is rclcpp/connext there is a 3 second delay most of the time.
* Test failure
  * The QoS on the c++ publisher is keep_last, depth 5, reliable,volatile. Subscriber QoS is keep_last, depth 10, reliable, volatile. The publisher publishes in bursts of 5 messages with a 1/100th of a second between the start of each message published. It does a burst 10 times per second. The python subscriber receives about 10 messages in the time it takes the publisher to publish 20 messages. The python subscriber is taking messages very close to half the rate that the publisher is publishing so I think something like this is happening (not exactly since subscriber taking is interleaved with publisher publishing):
     1. DataReader queue to fills up [1, 2, 3, 4, 5, 1, 2, 3, 4, 5].
     1. subscriber takes 2 messages: [3, 4, 5, 1, 2, 3, 4, 5, -, -]
     1. publisher sends another 5 messages, but data reader has to discard oldest 3: [1, 2, 3, 4, 5, 1, 2, 3, 4, 5]
     1. subscriber takes 3 messages: [4, 5, 1, 2, 3, 4, 5, -, -, -]
     1. publisher sends another 5 messages, but data reader has to discard oldest 2: [1, 2, 3, 4, 5, 1, 2, 3, 4, 5]
     1. Repeat until test times out (15 seconds total test time)

Recommendations for next steps:
* Startup delay
  * Wireshark the configurations where there is a consistent delay before subscriber receives the first message (rclcpp, multi-vendor)
* Test failure
  * Option 1:
    * print statements lower level to confirm if test failure really is due to timing and QoS settings
    * if so, profile and attempt to optimize rclpy to not be so slow on windows debug (it's not this slow using linux debug, what gives?)
  * Option 2:
    * Raise or lower the publish rate to increase the chance the subscriber takes message 5.

Experiments done while investigating the test failure

1. changed c++ publisher rate
    *  5 messages at 100hz 25 times per second
       * Test passes 10/10 times
    *  5 messages at 100hz 15 times per second
       * Test passes 10/10 times
    *  5 messages at 100hz 11 times per second
       * Test failed 4 times, passed on 5th attempt
    * (**current setting in master**): 5 messages@100hz10 times per second
      * Test fails 10/10 times
    *  5 messages at 100hz 9 times per second
       * Test fails 10/10 times
    *  5 messages at 100hz 8 times per second
       * Test passes 10/10 times
    *  5 messages at 100hz 5 times per second
       * Test passes 10/10 times
    *  5 messages at 100hz 1 time per second
       * Test passes 10/10 times

1. Changed subscriber queue depth
   * (**current setting in master**) depth 10
      * test fails 10/10 times
   * depth 9
      * test failed 5 times, passed on the 6th
   * depth 7
      * test failed 6 times, passed on the 7th
   * depth 5
      * test fails 10/10 times
   * depth 3
      * test fails twice, passes on 3rd attempt

Experiments done while investigating the startup delay.  Time measurements are inexact - it's me counting. Notes are **publisher** + **subscriber**.

1. connext/c++ + fast-rtps/python
   * 3 seconds between "beginning loop" from subscriber and subscriber printing that it RX'd a message
     * this delay is always about the same amount of time
   * test fails 10/10 times
1. added print statements and used connext/C++ + fast-rtps/python
   * executor waits in `rclpy_wait` immediately after "beginning loop" is printed, stays in there until a message is received (3 seconds)
   * test fails 10/10 times
1. connext/c++ + fast-rtps/c++
   * publisher publishes for 3 seconds before subscriber prints that it RX's the first message
     * this delay is always about the same amount of time
   * subscriber RX'd all 5 messages very quickly, once it starts (within 3 messages published in one publisher cycle, so probably ~0.04s for subscriber to RX all 5 messages)
   * test passes 10/10 times
1. connext/python + fast-rtps/python
   * zero delay between Subscriber starting and receiving the first message
   * Subscriber receives all messages as soon as publisher publishes them
   * test passes 10/10times
1. connext/python + fast-rtps/c++
   * 3/10 tests: there is a 3 second delay between when the publisher starts publishing messages and the subscriber receives it's first one
     * this delay is always about the same amount of time
   * 7/10 tests: No delay between start of publishing and start of subscriber receiving them.
   * test passes 10/10 times
1. fast-rtps/c++ + fast-rtps/c++
    *  Very short delay between publisher starting publishing and subscriber receiving first message (1-10 publisher cycles, so ~0.1s - 1s)
    * test passes 10/10 times
1. connext/c++ + connext/c++
    * variable delay between publisher publishing and subscriber receiving first message (0-3s, mostly closer to 3).
    * test passes 10/10 times
1. connext/python + connext/python
    * variable delay between publisher publishing and subscriber receiving first message (0-1s, mostly closer to 0).
    * test passes 10/10 times
1. fast-rtps/python + fast-rtps/python
    * variable delay between publisher publishing and subscriber receiving first message (0-0.5s, mostly closer to 0).
    * test passes 10/10 times
